src/OrleansRuntime/Core/InsideRuntimeClient.cs

```diff
@@ -414,6 +426,38 @@ public void SniffIncomingMessage(Message message)
             }
         }
 
+        private static void UnwrapCancellationToken(IAddressable target, object arg, InvokeMethodRequest request, int i)
+        {
+            // essentially unwrapping tokens that were wrapped before request sending
+            var orleansTokenWrapper = ((CancellationTokenWrapper)arg);
+            if (orleansTokenWrapper.WentThroughSerialization
```

Comment:
hopefully, once you decouple the serialization from the manager, `WentThroughSerialization` will go away, and this code will be simpler.